src/core/permissions/reducers/assignPermissionsToDataSetReducer.js

```diff
@@ -2,22 +2,75 @@
  * @flow
  */
 
-import { Map } from 'immutable';
+import { List, Map, Set } from 'immutable';
+import { Models } from 'lattice';
 import { RequestStates } from 'redux-reqseq';
 import type { SequenceAction } from 'redux-reqseq';
+import type {
+  Ace,
+  PermissionType,
+  UUID,
+} from 'lattice';
 
-import { REQUEST_STATE } from '../../redux/constants';
+import { ACES, ERROR, REQUEST_STATE } from '../../redux/constants';
 import {
   ASSIGN_PERMISSIONS_TO_DATA_SET,
   assignPermissionsToDataSet,
 } from '../actions';
 
+const { AceBuilder } = Models;
+
 export default function reducer(state :Map, action :SequenceAction) {
 
   // TODO: update stored aces
   return assignPermissionsToDataSet.reducer(state, action, {
-    REQUEST: () => state.setIn([ASSIGN_PERMISSIONS_TO_DATA_SET, REQUEST_STATE], RequestStates.PENDING),
-    SUCCESS: () => state.setIn([ASSIGN_PERMISSIONS_TO_DATA_SET, REQUEST_STATE], RequestStates.SUCCESS),
-    FAILURE: () => state.setIn([ASSIGN_PERMISSIONS_TO_DATA_SET, REQUEST_STATE], RequestStates.FAILURE),
+    REQUEST: () => state
+      .setIn([ASSIGN_PERMISSIONS_TO_DATA_SET, REQUEST_STATE], RequestStates.PENDING)
+      .setIn([ASSIGN_PERMISSIONS_TO_DATA_SET, action.id], action),
+    SUCCESS: () => {
+      const { permissions } = action.value;
+      const aces = state.get(ACES).withMutations((mutableAces :Map<List<UUID>, List<Ace>>) => {
+        permissions.forEach((ace :Ace, key :List<UUID>) => {
+          mutableAces.update(key, (storedAces :List<Ace>) => {
+
+            const targetIndex :number = storedAces.findIndex((storedAce :Ace) => (
+              storedAce.principal.id === ace.principal.id && storedAce.principal.type === ace.principal.type
+            ));
+
+            if (targetIndex >= 0) {
+              const targetAce :Ace = storedAces.get(targetIndex);
+
+              let updatedPermissions :Set<PermissionType> = Set(targetAce.permissions);
+              updatedPermissions = updatedPermissions.union(ace.permissions);
+
+              const updatedAce :Ace = (new AceBuilder())
+                .setPermissions(updatedPermissions)
+                .setPrincipal(targetAce.principal)
+                .build();
+
+              return storedAces.set(targetIndex, updatedAce);
+            }
+
+            if (targetIndex === -1) {
+              return storedAces.push(ace);
+            }
+
+            return storedAces;
+          });
+        });
```

Comment:
This block is a duplicate to what is in `updatePermissionsReducer`, except that it doesn't handle `ActionTypes.REMOVE`. I think it's a good idea to pull out this logic into a utility function and use it in both reducers, something like...

(this is just copy/paste from `updatePermissionsReducer`)
```
// src/core/permissions/reducers/assignPermissionsToDataSetReducer.js
const aces = state.get(ACES).withMutations((mutableAces :Map<List<UUID>, List<Ace>>) => {
  updatePermissionsInState(ActionTypes.ADD, permissions, mutableAces);
});

// src/core/permissions/reducers/updatePermissionsReducer.js
const aces = state.get(ACES).withMutations((mutableAces :Map<List<UUID>, List<Ace>>) => {
  updatePermissionsInState(actionType, permissions, mutableAces);
});

// src/core/permissions/utils/updatePermissionsInState.js
function updatePermissionsInState(
  actionType :ActionType,
  permissions :Map<List<UUID>, Ace>,
  stateAces :Map<List<UUID>, List<Ace>>,
) {

  permissions.forEach((ace :Ace, key :List<UUID>) => {
    stateAces.update(key, (storedAces :List<Ace>) => {

      const targetIndex :number = storedAces.findIndex((storedAce :Ace) => (
        storedAce.principal.id === ace.principal.id && storedAce.principal.type === ace.principal.type
      ));

      if (targetIndex >= 0) {
        const targetAce :Ace = storedAces.get(targetIndex);

        let updatedPermissions :Set<PermissionType> = Set(targetAce.permissions);
        if (actionType === ActionTypes.ADD) {
          updatedPermissions = updatedPermissions.union(ace.permissions);
        }
        else if (actionType === ActionTypes.REMOVE) {
          updatedPermissions = updatedPermissions.subtract(ace.permissions);
        }

        const updatedAce :Ace = (new AceBuilder())
          .setPermissions(updatedPermissions)
          .setPrincipal(targetAce.principal)
          .build();

        return storedAces.set(targetIndex, updatedAce);
      }

      if (targetIndex === -1 && actionType === ActionTypes.ADD) {
        return storedAces.push(ace);
      }

      return storedAces;
    });
  });
}
```